parser/auth/sm3.go

```diff
@@ -0,0 +1,387 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"strconv"
+)
+
+// The concrete SM3 Cryptographic Hash Algorithm can be accessed in http://www.sca.gov.cn/sca/xwdt/2010-12/17/content_1002389.shtml
+// This implementation of 'type sm3 struct' is modified from https://github.com/tjfoc/gmsm/tree/601ddb090dcf53d7951cc4dcc66276e2b817837c/sm3
+// Some other references:
+// 	https://tools.ietf.org/id/draft-oscca-cfrg-sm3-01.html
+
+/*
+Copyright Suzhou Tongji Fintech Research Institute 2017 All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+                 http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+type sm3 struct {
+	digest      [8]uint32 // digest represents the partial evaluation of V
+	length      uint64    // length of the message
+	unhandleMsg []byte
+	blockSize   int
+	size        int
+}
+
+func ff0(x, y, z uint32) uint32 { return x ^ y ^ z }
+
+func ff1(x, y, z uint32) uint32 { return (x & y) | (x & z) | (y & z) }
+
+func gg0(x, y, z uint32) uint32 { return x ^ y ^ z }
+
+func gg1(x, y, z uint32) uint32 { return (x & y) | (^x & z) }
+
+func p0(x uint32) uint32 { return x ^ leftRotate(x, 9) ^ leftRotate(x, 17) }
+
+func p1(x uint32) uint32 { return x ^ leftRotate(x, 15) ^ leftRotate(x, 23) }
+
+func leftRotate(x uint32, i uint32) uint32 { return x<<(i%32) | x>>(32-i%32) }
+
+func (sm3 *sm3) pad() []byte {
+	msg := sm3.unhandleMsg
+	// Append '1'
+	msg = append(msg, 0x80)
+	// Append until the resulting message length (in bits) is congruent to 448 (mod 512)
+	blockSize := 64
+	for i := len(msg); i%blockSize != 56; i++ {
+		msg = append(msg, 0x00)
+	}
+	// append message length
+	msg = append(msg, uint8(sm3.length>>56&0xff))
+	msg = append(msg, uint8(sm3.length>>48&0xff))
+	msg = append(msg, uint8(sm3.length>>40&0xff))
+	msg = append(msg, uint8(sm3.length>>32&0xff))
+	msg = append(msg, uint8(sm3.length>>24&0xff))
+	msg = append(msg, uint8(sm3.length>>16&0xff))
+	msg = append(msg, uint8(sm3.length>>8&0xff))
+	msg = append(msg, uint8(sm3.length>>0&0xff))
+
+	if len(msg)%64 != 0 {
+		panic("------sm3 Pad: error msgLen =")
```

Comment:
For arbitrary input, the code should never goes here?

Reply:
We should avoid user input make the server panic

Reply:
Yes, the `for` loop in the `pad()` function ensure the length of the final `msg` is a multiple of 64. Let me remove this `panic()`